### ISSUE TITLE
Vitest UI crashes server at startup

### ISSUE THREAD

**Issue by tony19** [vitest-dev/vitest#627]
### Describe the bug

Starting Vitest UI crashes the server at startup with this error:

```
node:internal/process/promises:246
          triggerUncaughtException(err, true /* fromPromise */);
          ^

[UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "#<Object>".] {
  code: 'ERR_UNHANDLED_REJECTION'
}
```

While this is consistently reproducible locally, this occurs *intermittently* on StackBlitz (via https://vitest.new), and I believe it causes StackBlitz's [preview pane to be a blank white screen](https://github.com/stackblitz/webcontainer-core/issues/536#issuecomment-1020712514) (due to the exception halting the rendering).

<details>
<summary>Console log</summary>

>     ➜  vite-test-tmp1 npm run test:ui
>     Debugger attached.
>     
>     > vite-test-tmp1@0.0.0 test:ui
>     > vitest --ui
>     
>     Debugger attached.
>     
>      WATCH  /Users/tony/src/tmp/vite-test-tmp1
>     
>      √ test/basic.test.ts (2)
>      √ test/suite.test.ts (3)
>     
>     Test Files  2 passed (2)
>          Tests  5 passed (5)
>           Time  3.42s (in thread 15ms, 23458.85%)
>     
>     
>      PASS  Waiting for file changes...
>            press h to show help, press q to quit
>     Waiting for the debugger to disconnect...
>     node:internal/process/promises:246
>               triggerUncaughtException(err, true /* fromPromise */);
>               ^
>     
>     [UnhandledPromiseRejection: This error originated either by throwing inside of an async function without a catch block, or by rejecting a promise which was not handled with .catch(). The promise rejected with the reason "#<Object>".] {
>       code: 'ERR_UNHANDLED_REJECTION'
>     }
>     
>     Node.js v17.0.1

</details>

Screenshot of debugger stopped at exception:

<img width="1619" alt="Screen Shot 2022-01-24 at 7 37 43 PM" src="https://user-images.githubusercontent.com/26580/150894769-5c4b0ded-3305-41d1-b466-854015435550.png">

#### Steps to reproduce

1. Setup repo (option 1):
  a. Clone https://github.com/tony19-sandbox/vitest-ui-crash-at-startup
  b. Install dependencies with `npm install`.

2. Setup repo (option 2):
  a. Scaffold a new project with `npm init vite -- --template vanilla`.
  b. Add `vitest` with `npm install -D vitest`.
  c. Add unit tests to `<projectRoot>/tests`.
  d. Add `test:ui` NPM script that runs `vitest --ui`.

3. Start `test:ui` with `npm run test:ui`.
4. Observe the error message in the terminal.

### System Info

```shell
System:
    OS: macOS 11.2.3
    CPU: (8) x64 Intel(R) Core(TM) i7-6820HQ CPU @ 2.70GHz
    Memory: 46.60 MB / 16.00 GB
    Shell: 5.8 - /bin/zsh
  Binaries:
    Node: 17.0.1 - ~/n/bin/node
    Yarn: 1.16.0 - ~/n/bin/yarn
    npm: 8.1.0 - ~/n/bin/npm
  Browsers:
    Chrome: 97.0.4692.71
    Chrome Canary: 99.0.4813.0
    Edge: 97.0.1072.69
    Firefox: 91.0.2
    Safari: 14.0.3
  npmPackages:
    vite: ^2.7.2 => 2.7.13 
    vitest: ^0.2.1 => 0.2.1 
```

### Used Package Manager

npm
yarn

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide).
- [X] Check that there isn't [already an issue](https://github.com/vitest-dev/vitest/issues) that reports the same bug to avoid creating a duplicate.
- [X] Check that this is a concrete bug. For Q&A open a [GitHub Discussion](https://github.com/vitest-dev/vitest/discussions) or join our [Discord Chat Server](https://chat.vitest.dev).
- [X] The provided reproduction is a [minimal reproducible example](https://stackoverflow.com/help/minimal-reproducible-example) of the bug.

**Comment by Demivan** [comment]
@tony19, should be fixed in v0.2.3

**Comment by bjarkihall** [comment]
According to discussions on discord, the `--ui` bug is due to the default port `51204` is reserved when starting the dev server, which you can work around by providing an unreserved api port (e.g. 3002) and it should work.

Maybe the server initialization is just missing a port check.

If no port is specified and the default port number is reserved, should it pick the next available port or throw an error indicating which port is blocked in a more readable way?

I remember a similar issue in the early days of SnowPack: https://github.com/withastro/snowpack/issues/386 and I thought vite had something to solve this too.

**Comment by sheremet-va** [comment]
> Maybe the server initialization is just missing a port check?
> 
> If no port is specified and the default port number is reserved, should it pick the next available port or throw an error indicating which port is blocked in a more readable way?

According to [Vite docs](https://vitejs.dev/config/#server-port), it does:

```
  /**
   * Specify server port. Note if the port is already being used, Vite will
   * automatically try the next available port so this may not be the actual
   * port the server ends up listening on.
   */
  port?: number;
```

Maybe it's because we are hardcoding port for hmr?

https://github.com/vitest-dev/vitest/blob/c23be109deb67f796bc31edb61d3acbb306fea49/packages/ui/client/composables/client.ts#L9

Haven't look what is actually does tho.

**Comment by edimitchel** [comment]
@tony19 can you confirm that this issue is still a problem for you? Feel free to reopen!

**Comment by tony19** [comment]
Confirmed fixed with 0.2.8. Thanks.
